### PR TITLE
add key-rotation for SSE-S3 objects

### LINKED ISSUE #6754: [SSE-S3]: add support for SSE-S3 object key rotation
*This is a feature request - therefore not using the template on purpose.*

The S3 protocol allows object key rotation for SSE-C. This means that S3 clients can specify the
two different SSE-C keys for an existing object during COPY (`dst == src`). AWS will verify that the 1st key is the correct current SSE-C key and than rotate the object key to the 2nd SSE-C key. (Simplified)

A similar mechanism should exists for SSE-S3 objects where the object key is protected by a CMK at the KMS. If the CMK is changed (replaced or new `key-id`) than a S3 client should be able to request a key-rotation similar to SSE-C. 

One important remark here is that key-rotation does **not** re-encrypt the object. The reason is that this would be a very expensive operation if objects are large. Instead key rotation replaces a key-encryption-key used to encrypted the actual secret object key.

### PR DESCRIPTION
**Pull request by aead**
## Description
This commit adds key-rotation for SSE-S3 objects.
To execute a key-rotation a SSE-S3 client must
 - specify the `X-Amz-Server-Side-Encryption: AES256` header
   for the destination
 - The source == destination for the COPY operation.

## Motivation and Context
Fixes #6754

## Regression
no

## How Has This Been Tested?
manually

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-tls.sh | :x: [more...](#9093bbdd-d10d-4a00-a25c-aba76dcd8752) |
<a name="9093bbdd-d10d-4a00-a25c-aba76dcd8752"></a>
### 6755-40cc414/mint-tls.sh.log:
```
Running with
SERVER_ENDPOINT:      192.168.39.64:9000
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         1
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 23db38429d05:/mint/log /tmp/mint-logs'

(1/7) Running aws-sdk-go tests ... done in 1 seconds
(2/7) Running aws-sdk-java tests ... done in 22 seconds
(3/7) Running mc tests ... done in 1 minutes and 1 seconds
(4/7) Running minio-dotnet tests ... done in 9 seconds
(5/7) Running minio-go tests ... FAILED in 21 seconds
{
  "args": {
    "destination": {},
    "source": {
      "Headers": {
        "X-Amz-Copy-Source": [
          "minio-go-test-qfzdyve3srjlzdz9/srcObject"
        ]
      }
    }
  },
  "duration": 359,
  "error": "Requests specifying Server Side Encryption with Customer provided keys must provide a valid encryption algorithm.",
  "function": "CopyObject(destination, source)",
  "message": "GetObject failed",
  "name": "minio-go: testEncryptedCopyObjectWrapper",
  "status": "FAIL"
}

Executed 4 out of 7 tests successfully.
```

**Comment by harshavardhana** [comment]
> (5/7) Running minio-go tests ... FAILED in 21 seconds { "args": { "destination": {}, "source": { "Headers": { "X-Amz-Copy-Source": [ "minio-go-test-qfzdyve3srjlzdz9/srcObject" ] } } }, "duration": 359, "error": "Requests specifying Server Side Encryption with Customer provided keys must provide a valid encryption algorithm.", "function": "CopyObject(destination, source)", "message": "GetObject failed", "name": "minio-go: testEncryptedCopyObjectWrapper", "status": "FAIL" }

Looks like the test failed @aead

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#fc194241-a8a8-4469-a02e-9cf81b4db126) |
<a name="fc194241-a8a8-4469-a02e-9cf81b4db126"></a>
### 6755-03befb4/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30143
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7dc9af812879:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 1 minutes and 27 seconds
(2/13) Running aws-sdk-go tests ... done in 1 seconds
(3/13) Running aws-sdk-java tests ... done in 5 seconds
(4/13) Running aws-sdk-php tests ... done in 50 seconds
(5/13) Running aws-sdk-ruby tests ... done in 6 seconds
(6/13) Running mc tests ... done in 53 seconds
(7/13) Running minio-dotnet tests ... FAILED in 3 minutes and 52 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.)) ---> System.AggregateException: One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.) ---> Minio.Exceptions.MinioException: Minio API responded with message=We encountered an internal error, please try again.
   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 363
   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 505
   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 351
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 483
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 229
   at Minio.Functional.Tests.FunctionalTest.PutObject_Task(MinioClient minio, String bucketName, String objectName, String fileName, String contentType, Int64 size, Dictionary`2 metaData, MemoryStream mstream) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 740
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.RemoveObjects_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1738
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 218

Executed 6 out of 13 tests successfully.
```

File: cmd/object-handlers.go [hunk 1]
```diff
@@ -829,26 +829,29 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
 			}
+
+			// In case of SSE-S3 oldKey and newKey aren't used - the KMS manages the keys.
 			if err = rotateKey(oldKey, newKey, srcBucket, srcObject, encMetadata); err != nil {
 				writeErrorResponse(w, toAPIErrorCode(err), r.URL)
 				return
 			}
-
-			// Since we are rotating the keys, make sure to update the metadata.
-			srcInfo.metadataOnly = true
```
Comment (harshavardhana) [inline cmd/object-handlers.go]: No need to change this - moving comment here is unrelated to your change.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/object-handlers.go (hunk 1 above)
Comment (aead) [reply cmd/object-handlers.go]: Auto-formating, done

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by poornas — CHANGES_REQUESTED** [review]
@aead, this check will also need to be changed. 
```
if !isMetadataReplace(r.Header) && srcInfo.metadataOnly && !crypto.SSEC.IsEncrypted(srcInfo.UserDefined) {
		// If x-amz-metadata-directive is not set to REPLACE then we need
		// to error out if source and destination are same.
		writeErrorResponse(w, ErrInvalidCopyDest, r.URL)
		return
	}
```
The copy operation below fails with minio but is ok with aws cli
```
aws s3api copy-object --bucket tbucket11 --copy-source tbucket11/brotate --key brotate --server-side-encryption AES256    --endpoint-url https://localhost:9000 --no-verify-ssl
```

**Comment by aead** [comment]
@poornas Oh - you're absolutely right. Fixed.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by aead** [comment]
CI:
```
--- FAIL: TestTreeWalkIsEnd (0.02s)
	tree-walk_test.go:642: remove C:\Users\travis\AppData\Local\Temp\minio-305829702\testvolume\i: The directory is not empty.
```

**Review by harshavardhana — APPROVED** [review]
LGTM

**Review by poornas — APPROVED** [review]
LGTM

**Pull request merged**